Avoid using generated client for federation types and use generic client instead

shashidharatd opened the pull request:
This pr builds on top of https://github.com/kubernetes-sigs/federation-v2/pull/602 and removes the usage of generated typed client for federation types. The usage is replaced by generic client from `sigs.k8s.io/controller-runtime` which will eventually aid us in upgrading to newer `kubebuilder` and also facilitate implementing leader election for federation controller.

/cc @marun

marun commented on pkg/controller/dnsendpoint/controller.go:
```diff
@@ -76,19 +77,21 @@ func newDNSEndpointController(client fedclientset.Interface, objectType pkgrunti
 		objectType,
 		util.NoResyncPeriod,
-		util.NewTriggerOnAllChanges(func(obj pkgruntime.Object) {
+		func(obj pkgruntime.Object) {
```
It's not necessary to pass a wrapper - just provide `d.enqueueObject` directly.

marun commented on pkg/controller/dnsendpoint/controller.go:
```diff
@@ -193,37 +198,43 @@ func (d *controller) processItem(key string) error {
-	if err != nil {
-		return err
-	}
+		err = d.client.Get(context.TODO(), objectKey, dnsEndpointObject)
```
It isn't necessary to retrieve an object to delete it.  Setting the name and namespace in the object meta of an object of the target type (e.g. DNSEndpoint in this case) is sufficient.

Reference: [Examples in the controller-runtime repo](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/example_test.go#L177)

marun commented on pkg/controller/dnsendpoint/ingress.go:
```diff
@@ -19,29 +19,22 @@ package dnsendpoint
-	}
-	watchFunc := func(options metav1.ListOptions) (watch.Interface, error) {
-		return client.MulticlusterdnsV1alpha1().IngressDNSRecords(config.TargetNamespace).Watch(options)
+	client, err := client.New(config.KubeConfig, client.Options{})
```
Consider initializing the client closer to where it is used - in `newDNSEndpointController` - since there is no longer any other use in this method.

marun commented on pkg/controller/dnsendpoint/service.go:
```diff
@@ -21,29 +21,22 @@ import (
-	}
-	watchFunc := func(options metav1.ListOptions) (watch.Interface, error) {
-		return client.MulticlusterdnsV1alpha1().ServiceDNSRecords(config.TargetNamespace).Watch(options)
+	client, err := client.New(config.KubeConfig, client.Options{})
```
Consider initializing the client closer to where it is used.

marun commented on pkg/controller/federatedcluster/controller.go:
```diff
@@ -78,15 +76,19 @@ type ClusterController struct {
+func StartClusterController(config *util.ControllerConfig, stopChan <-chan struct{}, clusterMonitorPeriod time.Duration) error {
 	fedClient, kubeClient, crClient := config.AllClients("cluster-controller")
-	controller := newClusterController(fedClient, kubeClient, crClient, config.FederationNamespaces, clusterMonitorPeriod)
+	controller, err := newClusterController(config, fedClient, kubeClient, crClient, clusterMonitorPeriod)
```
If config is passed in, is there a reason not to invoke `AllClients` in `newClusterController`?

marun commented on pkg/controller/federatedcluster/controller.go:
```diff
@@ -78,15 +76,19 @@ type ClusterController struct {
 
 // newClusterController returns a new cluster controller
-func newClusterController(fedClient fedclientset.Interface, kubeClient kubeclientset.Interface, crClient crclientset.Interface, namespaces util.FederationNamespaces, clusterMonitorPeriod time.Duration) *ClusterController {
+func newClusterController(config *util.ControllerConfig, fedClient client.Client, kubeClient kubeclientset.Interface, crClient crclientset.Interface, clusterMonitorPeriod time.Duration) (*ClusterController, error) {
```
(No action required) The generic client could replace not just usage of the generated fed client, but also the kube and cr clients.  Not saying it has to happen in this PR, mind you.

Also, `fedClient` suggests a federation-specific client, but `client.Client` can be used to access any api resource, so an eventual rename is suggested.

shashidharatd replied:
> The generic client could replace not just usage of the generated fed client, but also the kube and cr clients. Not saying it has to happen in this PR, mind you.

yes, it should be in subsequent pr's. Already this pr is big.

> Also, `fedClient` suggests a federation-specific client, but `client.Client` can be used to access any api resource, so an eventual rename is suggested.

Renamed the variable properly, keeping future in mind.

marun commented on pkg/controller/federatedtypeconfig/controller.go:
```diff
@@ -70,7 +70,10 @@ func StartController(config *util.ControllerConfig, stopChan <-chan struct{}) er
 	kubeConfig := config.KubeConfig
 	restclient.AddUserAgent(kubeConfig, userAgent)
-	client := fedclientset.NewForConfigOrDie(kubeConfig).CoreV1alpha1()
+	client, err := client.New(kubeConfig, client.Options{})
```
(No action required) Consider choosing a variable name that is differentiated from the package name, either by aliasing the import (e.g. `genclient`) or naming the variable differently (e.g. genClient).

marun commented on pkg/controller/schedulingmanager/controller.go:
```diff
@@ -59,43 +54,39 @@ func StartSchedulerController(config *util.ControllerConfig, stopChan <-chan str
-	controller := newController(config, client)
+	controller, err := newController(config)
+	if err != nil {
+		panic(err)
```
Why is a panic suggested here vs bubbling up?

marun commented on pkg/controller/schedulingpreference/controller.go:
```diff
@@ -76,8 +72,8 @@ type SchedulingPreferenceController struct {
 	userAgent := fmt.Sprintf("%s-controller", schedulingType.Kind)
-	fedClient, kubeClient, crClient := config.AllClients(userAgent)
-	controller, err := newSchedulingPreferenceController(config, schedulingType.SchedulerFactory, fedClient, kubeClient, crClient)
+	_, kubeClient, _ := config.AllClients(userAgent)
```
(No action required) Is this a hint that it should be possible to initialize just the generic client from the controller config?

marun commented on pkg/controller/servicedns/controller.go:
```diff
@@ -358,7 +354,7 @@ func (c *Controller) reconcile(qualifiedName util.QualifiedName) util.Reconcilia
 
 	if !reflect.DeepEqual(cachedDNS.Status, fedDNS.Status) {
-		_, err = c.fedClient.MulticlusterdnsV1alpha1().ServiceDNSRecords(fedDNS.Namespace).UpdateStatus(fedDNS)
+		err = c.fedClient.Status().Update(context.TODO(), fedDNS)
```
Why is `context.TODO` preferred over `context.Background`?

shashidharatd replied:
while both are same, context.TODO reminds us that it needs a rework in the future to use the single context for the code path across the functions. it is not logical to use context.Background() everywhere. here is the comment on its usage:
```
// Background returns a non-nil, empty Context. It is never canceled, has no
// values, and has no deadline. It is typically used by the main function,
// initialization, and tests, and as the top-level Context for incoming
// requests.
```
```
// TODO returns a non-nil, empty Context. Code should use context.TODO when
// it's unclear which Context to use or it is not yet available (because the
// surrounding function has not yet been extended to accept a Context
// parameter). TODO is recognized by static analysis tools that determine
// whether Contexts are propagated correctly in a program.
```

marun commented on pkg/controller/status/controller.go:
```diff
@@ -91,32 +92,32 @@ func StartFederationStatusController(controllerConfig *util.ControllerConfig, st
 
 // newFederationStatusController returns a new status controller for the federated type
-func newFederationStatusController(controllerConfig *util.ControllerConfig, typeConfig typeconfig.Interface) (*FederationStatusController, error) {
+func newFederationStatusController(config *util.ControllerConfig, typeConfig typeconfig.Interface) (*FederationStatusController, error) {
```
(No action required) What is your thinking behind s/controllerConfig/config/?  Config is such a generic term that I used `controllerConfig` to eliminate potential ambiguity.

shashidharatd replied:
Was trying to be coherent. I know these changes should be ideally separate as you said. would try to conform in future.

marun commented on pkg/controller/sync/version/cluster.go:
```diff
@@ -65,18 +71,28 @@ func (a *clusterVersionAdapter) SetStatus(obj pkgruntime.Object, status *fedv1a1
-
-func (a *clusterVersionAdapter) Delete(qualifiedName util.QualifiedName) error {
-	return a.client.ClusterPropagatedVersions().Delete(qualifiedName.Name, nil)
+	objectKey := client.ObjectKey{Name: qualifiedName.Name}
```
(No action required) I think `QualifiedName` is useful in the way it supports stringifying cleanly for both namespaced and non-namespaced resources.   Maybe it makes sense to add methods and helpers that convert to `client.ObjectKey` since that will be used in all the client calls?

marun commented on pkg/controller/util/federated_informer.go:
```diff
@@ -176,79 +173,78 @@ func NewFederatedInformer(
+	}
+
+	var err error
+	federatedInformer.clusterInformer.store, federatedInformer.clusterInformer.controller, err = NewGenericInformerWithEventHandler(
```
Why is it desirable to change the code to accept the event handler functions separately?

shashidharatd replied:
reverted. ptal

marun commented on pkg/controller/schedulingmanager/controller.go:
```diff
@@ -20,21 +20,16 @@ import (
 	"github.com/golang/glog"
 	"github.com/pkg/errors"
 
+	"k8s.io/apimachinery/pkg/util/runtime"
```
(No action required) I'm not going to insist in this instance since the PR is large and I'm not sure how easy it is for you to split it apart (I have tools that make it easy), but in future please try to separate non-functional changes like import ordering into separate commits for a cleaner history.

shashidharatd replied:
got your point. would keep in mind in future pr's.

shashidharatd commented on test/e2e/servicedns.go:
```diff
@@ -43,27 +43,27 @@ var _ = Describe("ServiceDNS", func() {
 	const userAgent = "test-service-dns"
 	const baseName = "test-service-dns-"
-	const federation = "galactic"
+	const federationPrefix = "galactic"
```
Separated this code change into another commit

shashidharatd commented on pkg/controller/federatedcluster/controller.go:
```diff
@@ -78,43 +75,43 @@ type ClusterController struct {
-func StartClusterController(config *util.ControllerConfig, stopChan <-chan struct{}, clusterMonitorPeriod time.Duration) {
-	fedClient, kubeClient, crClient := config.AllClients("cluster-controller")
-	controller := newClusterController(fedClient, kubeClient, crClient, config.FederationNamespaces, clusterMonitorPeriod)
+func StartClusterController(controllerConfig *util.ControllerConfig, stopChan <-chan struct{}, clusterMonitorPeriod time.Duration) error {
```
Done.

shashidharatd commented on pkg/client/generic/genericclient.go:
```diff
@@ -0,0 +1,28 @@
+	"github.com/kubernetes-sigs/federation-v2/pkg/apis"
+)
+
+func New(config *rest.Config) (client.Client, error) {
```
Done.

shashidharatd commented on pkg/controller/federatedtypeconfig/controller.go:
```diff
@@ -66,13 +67,16 @@ type Controller struct {
 
 // StartController starts the Controller for managing FederatedTypeConfig objects.
-func StartController(config *util.ControllerConfig, stopChan <-chan struct{}) error {
+func StartController(controllerConfig *util.ControllerConfig, stopChan <-chan struct{}) error {
```
Done.

shashidharatd commented on pkg/client/generic/genericclient.go:
```diff
@@ -0,0 +1,28 @@
+)
+
+func New(config *rest.Config) (client.Client, error) {
+	return client.New(config, client.Options{Scheme: apis.Scheme})
```
If we didn't provide the populated scheme, the clients fails for federation api, like the below error:
```
Error creating FederatedTypeConfig "clusterroles.rbac.authorization.k8s.io": no kind is registered for the type v1alpha1.FederatedTypeConfig in scheme "k8s.io/client-go/kubernetes/scheme/register.go:60"
```

marun replied:
Interesting, I guess I must not have been triggering that failure in my previous use of the generic client.  Thank you for the clarification.

shashidharatd commented on pkg/controller/dnsendpoint/controller.go:
```diff
@@ -66,29 +67,33 @@ type controller struct {
 
-func newDNSEndpointController(client fedclientset.Interface, objectType pkgruntime.Object, objectKind string,
-	listFunc cache.ListFunc, watchFunc cache.WatchFunc, getEndpoints GetEndpointsFunc, minimizeLatency bool) (*controller, error) {
+func newDNSEndpointController(controllerConfig *util.ControllerConfig, objectType pkgruntime.Object, objectKind string,
```
Done.

shashidharatd commented on pkg/controller/dnsendpoint/controller.go:
```diff
@@ -193,37 +198,48 @@ func (d *controller) processItem(key string) error {
+			},
+		}
+		err = d.genClient.Delete(context.TODO(), dnsEndpointObject)
+		if err != nil && apierrors.IsNotFound(err) {
```
Done

shashidharatd commented on pkg/controller/dnsendpoint/controller.go:
```diff
@@ -193,37 +198,48 @@ func (d *controller) processItem(key string) error {
+		dnsEndpointObject := &feddnsv1a1.DNSEndpoint{}
+		objectKey := client.ObjectKey{Namespace: namespace, Name: name}
+		err = d.genClient.Get(context.TODO(), objectKey, dnsEndpointObject)
+		if err != nil {
```
Slightly refactored. ptal

shashidharatd commented on pkg/controller/dnsendpoint/controller.go:
```diff
@@ -193,37 +198,48 @@ func (d *controller) processItem(key string) error {
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				newDNSEndpointObject := &feddnsv1a1.DNSEndpoint{
+					ObjectMeta: metav1.ObjectMeta{
```
Not necessary now in the new code

shashidharatd commented on pkg/controller/dnsendpoint/controller.go:
```diff
@@ -193,37 +198,48 @@ func (d *controller) processItem(key string) error {
+		// Update only if the new endpoints are not equal to the existing ones.
+		if !reflect.DeepEqual(dnsEndpointObject.Spec.Endpoints, dnsEndpoints) {
+			dnsEndpointObject.Spec.Endpoints = dnsEndpoints
+			err = d.genClient.Update(context.TODO(), dnsEndpointObject)
```
Done

shashidharatd commented on pkg/controller/dnsendpoint/ingress.go:
```diff
@@ -19,29 +19,16 @@ package dnsendpoint
-
-	controller, err := newDNSEndpointController(client, &feddnsv1a1.IngressDNSRecord{}, "ingress",
-		listFunc, watchFunc, getIngressDNSEndpoints, config.MinimizeLatency)
+func StartIngressDNSEndpointController(controllerConfig *util.ControllerConfig, stopChan <-chan struct{}) error {
```
Done

shashidharatd commented on pkg/controller/dnsendpoint/service.go:
```diff
@@ -21,29 +21,16 @@ import (
-
-	controller, err := newDNSEndpointController(client, &feddnsv1a1.ServiceDNSRecord{}, "service",
-		listFunc, watchFunc, getServiceDNSEndpoints, config.MinimizeLatency)
+func StartServiceDNSEndpointController(controllerConfig *util.ControllerConfig, stopChan <-chan struct{}) error {
```
Done

shashidharatd commented on pkg/controller/federatedcluster/controller.go:
```diff
@@ -78,43 +75,43 @@ type ClusterController struct {
-			DeleteFunc: cc.delFromClusterSet,
+		&cache.ResourceEventHandlerFuncs{
 			AddFunc:    cc.addToClusterSet,
+			DeleteFunc: cc.delFromClusterSet,
```
Done

shashidharatd commented on pkg/controller/federatedtypeconfig/controller.go:
```diff
@@ -66,13 +67,16 @@ type Controller struct {
+	kubeConfig := controllerConfig.KubeConfig
 	restclient.AddUserAgent(kubeConfig, userAgent)
-	client := fedclientset.NewForConfigOrDie(kubeConfig).CoreV1alpha1()
+	genClient, err := genericclient.New(kubeConfig)
```
Done

shashidharatd commented on pkg/controller/federatedtypeconfig/controller.go:
```diff
@@ -82,10 +86,10 @@ func StartController(config *util.ControllerConfig, stopChan <-chan struct{}) er
 
 // newController returns a new controller to manage FederatedTypeConfig objects.
-func newController(config *util.ControllerConfig, client corev1alpha1client.CoreV1alpha1Interface) (*Controller, error) {
+func newController(controllerConfig *util.ControllerConfig, client client.Client) (*Controller, error) {
```
Done

shashidharatd commented on pkg/controller/ingressdns/controller.go:
```diff
@@ -70,13 +70,13 @@ type Controller struct {
-func StartController(config *util.ControllerConfig, stopChan <-chan struct{}) error {
-	fedClient, kubeClient, crClient := config.AllClients("IngressDNS")
-	controller, err := newController(config, fedClient, kubeClient, crClient)
+func StartController(controllerConfig *util.ControllerConfig, stopChan <-chan struct{}) error {
```
Done

shashidharatd commented on pkg/controller/ingressdns/controller.go:
```diff
@@ -85,11 +85,11 @@ func StartController(config *util.ControllerConfig, stopChan <-chan struct{}) er
 
 // newController returns a new controller to manage IngressDNSRecord objects.
-func newController(config *util.ControllerConfig, fedClient fedclientset.Interface, kubeClient kubeclientset.Interface, crClient crclientset.Interface) (*Controller, error) {
+func newController(controllerConfig *util.ControllerConfig, genClient client.Client, kubeClient kubeclientset.Interface, crClient crclientset.Interface) (*Controller, error) {
```
Done

shashidharatd commented on pkg/controller/ingressdns/controller.go:
```diff
@@ -101,28 +101,25 @@ func newController(config *util.ControllerConfig, fedClient fedclientset.Interfa
 		&dnsv1a1.IngressDNSRecord{},
 		util.NoResyncPeriod,
-		util.NewTriggerOnAllChanges(func(obj pkgruntime.Object) {
+		func(obj pkgruntime.Object) {
```
Removed.

shashidharatd commented on pkg/controller/schedulingmanager/controller.go:
```diff
@@ -47,55 +42,52 @@ type SchedulerController struct {
 	scheduler map[string]schedulingtypes.Scheduler
 
-	config *util.ControllerConfig
+	controllerConfig *util.ControllerConfig
```
Done

shashidharatd commented on pkg/controller/schedulingmanager/controller.go:
```diff
@@ -47,55 +42,52 @@ type SchedulerController struct {
 }
 
-func StartSchedulerController(config *util.ControllerConfig, stopChan <-chan struct{}) {
+func StartSchedulerController(controllerConfig *util.ControllerConfig, stopChan <-chan struct{}) error {
```
Done

shashidharatd commented on pkg/controller/schedulingmanager/controller.go:
```diff
@@ -47,55 +42,52 @@ type SchedulerController struct {
 }
 
-func newController(config *util.ControllerConfig, client corev1alpha1client.CoreV1alpha1Interface) *SchedulerController {
+func newController(controllerConfig *util.ControllerConfig) (*SchedulerController, error) {
```
Done

shashidharatd commented on pkg/controller/schedulingpreference/controller.go:
```diff
@@ -74,14 +71,12 @@ type SchedulingPreferenceController struct {
 }
 
 // SchedulingPreferenceController starts a new controller for given type of SchedulingPreferences
-func StartSchedulingPreferenceController(config *util.ControllerConfig, schedulingType schedulingtypes.SchedulingType) (schedulingtypes.Scheduler, error) {
```
Done

shashidharatd commented on pkg/controller/schedulingpreference/controller.go:
```diff
@@ -90,14 +85,21 @@ func StartSchedulingPreferenceController(config *util.ControllerConfig, scheduli
 
 // newSchedulingPreferenceController returns a new SchedulingPreference Controller for the given type
-func newSchedulingPreferenceController(config *util.ControllerConfig, schedulerFactory schedulingtypes.SchedulerFactory, fedClient fedclientset.Interface, kubeClient kubeclientset.Interface, crClient crclientset.Interface) (*SchedulingPreferenceController, error) {
+func newSchedulingPreferenceController(controllerConfig *util.ControllerConfig, schedulingType schedulingtypes.SchedulingType) (*SchedulingPreferenceController, error) {
```
Done

shashidharatd commented on pkg/controller/servicedns/controller.go:
```diff
@@ -79,13 +77,12 @@ type Controller struct {
-func StartController(config *util.ControllerConfig, stopChan <-chan struct{}) error {
-	fedClient, kubeClient, crClient := config.AllClients("ServiceDNS")
-	controller, err := newController(config, fedClient, kubeClient, crClient)
+func StartController(controllerConfig *util.ControllerConfig, stopChan <-chan struct{}) error {
```
Done

shashidharatd commented on pkg/controller/servicedns/controller.go:
```diff
@@ -94,13 +91,14 @@ func StartController(config *util.ControllerConfig, stopChan <-chan struct{}) er
 
 // newController returns a new controller to manage ServiceDNSRecord objects.
-func newController(config *util.ControllerConfig, fedClient fedclientset.Interface, kubeClient kubeclientset.Interface, crClient crclientset.Interface) (*Controller, error) {
+func newController(controllerConfig *util.ControllerConfig) (*Controller, error) {
```
Done

shashidharatd commented on pkg/controller/sync/version/cluster.go:
```diff
@@ -65,18 +71,27 @@ func (a *clusterVersionAdapter) SetStatus(obj pkgruntime.Object, status *fedv1a1
 	version := obj.(*fedv1a1.ClusterPropagatedVersion)
-	return a.client.ClusterPropagatedVersions().Create(version)
+	err := a.genClient.Create(context.TODO(), version)
+	if err != nil {
```
Have defined the new interface for generic client and used it now. ptal.

shashidharatd commented on pkg/controller/sync/version/cluster.go:
```diff
@@ -65,18 +71,27 @@ func (a *clusterVersionAdapter) SetStatus(obj pkgruntime.Object, status *fedv1a1
-func (a *clusterVersionAdapter) Delete(qualifiedName util.QualifiedName) error {
-	return a.client.ClusterPropagatedVersions().Delete(qualifiedName.Name, nil)
+	clusterPropagatedVersion := &fedv1a1.ClusterPropagatedVersion{}
+	err := a.genClient.Get(context.TODO(), qualifiedName.ObjectKey(), clusterPropagatedVersion)
```
Not necessary now, so removed.

shashidharatd commented on pkg/kubefed2/join.go:
```diff
@@ -17,29 +17,32 @@ limitations under the License.
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/wait"
-	client "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes"
```
Done

shashidharatd commented on pkg/kubefed2/unjoin.go:
```diff
@@ -14,24 +14,28 @@ limitations under the License.
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	client "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes"
```
Done

shashidharatd commented on pkg/kubefed2/util/util.go:
```diff
@@ -19,11 +19,12 @@ package util
-	fedclient "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
-	client "k8s.io/client-go/kubernetes"
+	genericclient "github.com/kubernetes-sigs/federation-v2/pkg/client/generic"
+	"k8s.io/client-go/kubernetes"
```
Done

shashidharatd commented on test/common/crudtester.go:
```diff
@@ -67,11 +69,16 @@ type TestCluster struct {
 }
 
 func NewFederatedTypeCrudTester(testLogger TestLogger, typeConfig typeconfig.Interface, kubeConfig *rest.Config, testClusters map[string]TestCluster, waitInterval, clusterWaitTimeout time.Duration) (*FederatedTypeCrudTester, error) {
+	genClient, err := genericclient.New(kubeConfig)
```
Done

shashidharatd commented on test/common/util.go:
```diff
@@ -24,3 +30,11 @@ type TestLogger interface {
 	Logf(format string, args ...interface{})
 }
+
+func NewClientForConfigOrDie(config *rest.Config) client.Client {
```
removed, and instead used the one defined in `pkg/client/generic`

shashidharatd commented on test/e2e/framework/framework.go:
```diff
@@ -52,7 +53,7 @@ type FederationFrameworkImpl interface {
 
 	KubeClient(userAgent string) kubeclientset.Interface
-	FedClient(userAgent string) fedclientset.Interface
+	GenClient(userAgent string) client.Client
```
used `Client` now

shashidharatd commented on test/e2e/framework/managed/federation.go:
```diff
@@ -208,10 +208,11 @@ func (f *FederationFixture) createSecret(tl common.TestLogger, clusterFixture *K
 func (f *FederationFixture) createFederatedCluster(tl common.TestLogger, clusterName, secretName string) {
-	fedClient := f.NewFedClient(tl, userAgent)
-	_, err := fedClient.CoreV1alpha1().FederatedClusters(f.SystemNamespace).Create(&fedv1a1.FederatedCluster{
+	genClient := f.NewFedClient(tl, userAgent)
```
Done

shashidharatd commented on test/e2e/servicedns.go:
```diff
@@ -43,27 +43,27 @@ var _ = Describe("ServiceDNS", func() {
-		return dnsClient.Get(name, metav1.GetOptions{})
+		objectKey := client.ObjectKey{Namespace: namespace, Name: name}
+		serviceDNSRecords := &dnsv1a1.ServiceDNSRecord{}
+		err := genClient.Get(context.TODO(), objectKey, serviceDNSRecords)
```
Done

shashidharatd commented on test/e2e/version.go:
```diff
@@ -66,17 +66,27 @@ type testVersionAdapter interface {
+			Name:      qualifiedName.Name,
+		},
+	}
+	return a.coreClient.Delete(context.TODO(), propagatedVersion)
```
Done

shashidharatd commented on test/e2e/version.go:
```diff
@@ -178,9 +197,9 @@ func (r *testVersionedResource) OverrideVersion() (string, error) {
-	coreClient := fedClient.CoreV1alpha1()
+func newTestVersionAdapter(genClient client.Client, kubeClient kubeclientset.Interface, namespaced bool) testVersionAdapter {
+	adapter := version.NewVersionAdapter(genClient, namespaced)
+	coreClient := genClient
```
Done

shashidharatd commented:
@marun, Thanks for your detailed review. Have addressed all comments. PTAL. Thanks !
Also i am fine to remove the auto-generated tests for api (which use generated clients).
Do  let me know if i should remove them in the current pr.
Thanks !

marun commented on pkg/kubefed2/join.go:
```diff
@@ -634,7 +640,7 @@ func createHealthCheckClusterRoleAndBinding(clientset client.Interface, saName,
 				NonResourceURLs: []string{"/healthz"},
 			},
-			// The cluster client expects to be able to list nodes to retrieve zone and region details.
+			// The cluster kubernetes expects to be able to list nodes to retrieve zone and region details.
```
nit: s/kubernetes/client/

shashidharatd replied:
Done

marun commented on test/e2e/framework/framework.go:
```diff
@@ -235,7 +234,7 @@ func (f *frameworkWrapper) EnsureTestNamespacePropagation() {
 	tl := f.Logger()
 
-	dynclient, err := client.New(f.KubeConfig(), client.Options{})
+	client, err := genericclient.New(f.KubeConfig())
```
(No action requried) This could use `f.Client`

marun commented on test/e2e/framework/framework.go:
```diff
@@ -272,30 +270,26 @@ func (f *frameworkWrapper) EnsureTestFederatedNamespace(allClusters bool) *unstr
+	err = client.Create(context.Background(), obj)
 	if err != nil {
-		tl.Fatalf("Error creating %s %q: %v", apiResource.Kind, key, err)
+		tl.Fatalf("Error creating %s %q: %v", apiResource.Kind, namespace, err)
```
A federated namespace is namespaced, so as repetitive as it may be the name will be `<namespace>/<namespace>`, not `<namespace>`.  If you don't like that, rewording as follows would work: 

`Error creating  <kind> for namespace <namespace>`

shashidharatd replied:
Done. used the latter. Thanks !

marun commented on test/e2e/framework/framework.go:
```diff
@@ -272,30 +270,26 @@ func (f *frameworkWrapper) EnsureTestFederatedNamespace(allClusters bool) *unstr
 	if f.namespaceTypeConfig == nil {
 		tl := f.Logger()
-		dynClient, err := client.New(f.KubeConfig(), client.Options{})
+		client, err := genericclient.New(f.KubeConfig())
```
(No action required) -> `f.Client`

shashidharatd commented on pkg/controller/sync/version/cluster.go:
```diff
@@ -17,29 +17,32 @@ limitations under the License.
 
 type clusterVersionAdapter struct {
-	client corev1alpha1.CoreV1alpha1Interface
+	client genericclient.Client
```
yeah, i had dilemma. when using Client interface, `genericclient` alias will lead to sth like `genericclient.Client` and in some places if i just used `generic` and use the New method, it leads to `generic.New`. So i ended up like the way it is.

marun replied:
Ok, thank you for the clarification.

shashidharatd commented on pkg/controller/dnsendpoint/controller.go:
```diff
@@ -193,37 +197,42 @@ func (d *controller) processItem(key string) error {
-		}
+	dnsEndpointObject := &feddnsv1a1.DNSEndpoint{}
+	err = d.client.Get(context.TODO(), dnsEndpointObject, namespace, name)
+	if err != nil && !apierrors.IsNotFound(err) {
```
Done

shashidharatd commented:
@Addressed nits. ptal. Thanks!

marun commented:
/lgtm

Thank you!

shashidharatd commented:
@marun, just rebased to resolve conflicts. could you ptal and lgtm again. Thanks!

marun commented:
@shashidharatd My apologies, I forgot this review was on hold.  I didn't mean to merge anything before this merged.

marun commented:
/lgtm

shashidharatd commented:
Thanks @marun
